// pageres/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _events = require('events');

var _arrayUniq = require('array-uniq');

var _arrayUniq2 = _interopRequireDefault(_arrayUniq);

var _arrayDiffer = require('array-differ');

var _arrayDiffer2 = _interopRequireDefault(_arrayDiffer);

var _objectAssign = require('object-assign');

var _objectAssign2 = _interopRequireDefault(_objectAssign);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class Pageres {
	/**
  * Initialize a new Pageres
  *
  * @param {Object} options
  * @api public
  */

	constructor(options) {
		this.options = (0, _objectAssign2.default)({}, options);
		this.options.filename = this.options.filename || '<%= url %>-<%= size %><%= crop %>';
		this.options.format = this.options.format || 'png';

		this.stats = {};
		this.items = [];
		this.sizes = [];
		this.urls = [];
	}

	/**
  * Get or set page to capture
  *
  * @param {String} url
  * @param {Array} sizes
  * @param {Object} options
  * @api public
  */

	src(url, sizes, options) {
		if (!arguments.length) {
			return this._src;
		}

		this._src = this._src || [];
		this._src.push({ url: url, sizes: sizes, options: options });

		return this;
	}

	/**
  * Get or set the destination directory
  *
  * @param {String} dir
  * @api public
  */

	dest(dir) {
		if (!arguments.length) {
			return this._dest;
		}

		this._dest = dir;
		return this;
	}

	/**
  * Run pageres
  *
  * @api public
  */

	run() {
		var _this = this;

		return (0, _asyncToGenerator3.default)(function* () {
			yield _promise2.default.all(_this.src().map(function (src) {
				const options = (0, _objectAssign2.default)({}, _this.options, src.options);
				const sizes = (0, _arrayUniq2.default)(src.sizes.filter(/./.test, /^\d{2,4}x\d{2,4}$/i));
				const keywords = (0, _arrayDiffer2.default)(src.sizes, sizes);

				if (!src.url) {
					throw new Error('URL required');
				}

				_this.urls.push(src.url);

				if (!sizes.length && keywords.indexOf('w3counter') !== -1) {
					return _this.resolution(src.url, options);
				}

				if (keywords.length) {
					return _this.viewport({ url: src.url, sizes: sizes, keywords: keywords }, options);
				}

				for (const size of sizes) {
					_this.sizes.push(size);
					_this.items.push(_this.create(src.url, size, options));
				}
			}));

			_this.stats.urls = (0, _arrayUniq2.default)(_this.urls).length;
			_this.stats.sizes = (0, _arrayUniq2.default)(_this.sizes).length;
			_this.stats.screenshots = _this.items.length;

			if (!_this.dest()) {
				return _this.items;
			}

			yield _this.save(_this.items);

			return _this.items;
		})();
	}
}

exports.default = Pageres;
(0, _objectAssign2.default)(Pageres.prototype, _events.EventEmitter.prototype);
(0, _objectAssign2.default)(Pageres.prototype, require('./util'));
module.exports = exports['default'];
